feat: Marcinkiewicz-Zygmund inequality

YaelDillies opened the pull request:
Prove the Marcinkiewicz-Zygmund inequality.

The Marcinkiewicz-Zygmund inequality states that, if `X₁, ..., Xₐ ∈ L^p` are independent random
variables of mean zero valued in some inner product space, then the `L^p`-norm of `X₁ + ... + Xₐ` is
at most `Cₚ` times the `L^(p/2)`-norm of `|X₁|² + ... + |Xₐ|²`, where `Cₚ` is a constant depending on
`p` alone.

From LeanAPAP

Co-authored-by: Bhavik Mehta <bhavikmehta8@gmail.com>

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->
- [x] depends on: #15832
- [x] depends on: #21033

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

YaelDillies commented on Mathlib/Analysis/Complex/MarcinkiewiczZygmund.lean:
```diff
@@ -0,0 +1,274 @@
+/-- The **Marcinkiewicz-Zygmund inequality** for real-valued functions, with a slightly easier to
+bound constant than `Real.marcinkiewicz_zygmund'`.
+
+Note that `RCLike.marcinkiewicz_zygmund` is another  -/
```
```suggestion
Note that `RCLike.marcinkiewicz_zygmund` is another version that works for both `ℝ` and `ℂ` at the
expense of a slightly worse constant. -/
```

b-mehta commented on Mathlib/Analysis/Complex/MarcinkiewiczZygmund.lean:
```diff
@@ -0,0 +1,275 @@
+      m ^ m * 2 ^ (m + (m - 1)) *
+        ((∑ i, f (a i) ^ 2) ^ m + (∑ i, f (b i) ^ 2) ^ m) := by
+  rw [pow_add, ← mul_assoc _ _ (2 ^ _), mul_assoc _ (2 ^ (m - 1)), sum_add_distrib]
+  gcongr
```
I know this isn't enforced, but I feel pretty strongly that `gcongr` should always have an explicit pattern to make it more robust. At some point I'll make a stronger case for this on zulip, but in the mean time would you mind squeezing the `gcongr`s in this PR?

YaelDillies replied:
I don't really fancy duplicating all the LHS or RHS every time I use `gcongr`, especially since all `gcongr`s here are maximal (in the sense that the goals it leaves are not of the form `r (f a) (f b)` where `r` is a concrete relation and `f` a concrete operation) so I will wait until you have convinced the rest of the community that this is worth doing

b-mehta replied:
Eh, we really should have `gcongr?`

YaelDillies commented on Mathlib/Analysis/Complex/MarcinkiewiczZygmund.lean:
```diff
@@ -0,0 +1,275 @@
+/-!
+# The Marcinkiewicz-Zygmund inequality
+
+This file proves the Marcinkiewicz-Zygmund inequality. This is the statement that, for a real- or
```
Yes we did

YaelDillies commented on Mathlib/Analysis/Complex/MarcinkiewiczZygmund.lean:
```diff
@@ -0,0 +1,275 @@
+-- feels like could generalise more...
+-- the key point is that you combine the double sums into a single sum, and do a pair swap
+-- when the corresponding ε is -1
+-- but the order here is a bit subtle (ie this explanation is an oversimplification)
```
You did. I would never have written that because I would have found the generalisation instead 😉

b-mehta replied:
Okay cool, it's nonsense, let's delete it. The generalisation given here doesn't work.

YaelDillies commented on Mathlib/Analysis/Complex/MarcinkiewiczZygmund.lean:
```diff
@@ -0,0 +1,275 @@
+
+namespace Real
+
+/-- The **Marcinkiewicz-Zygmund inequality** for real-valued functions, with a slightly better
```
I've added a todo in the module doc

b-mehta commented on Mathlib/Analysis/Complex/MarcinkiewiczZygmund.lean:
```diff
@@ -0,0 +1,248 @@
+## TODO
+
+We currently only prove the inequality for `p = 2 * m` an even natural number. The general `p` case
+can be obtained from this specific one by nesting of LP norms.
```
```suggestion
can be obtained from this specific one by nesting of Lp norms.
```

hrmacbeth commented:
Does this work for Euclidean space, by expressing the norm-square of a vector as the sum of the squares of its components?  You do this for the real+complex decomposition of a complex number, but I don't immediately see anything special to two summands.  (Of course, the constants would be adjusted.)

If that works, then you can extend to a finite-dimensional inner product space by using `stdOrthonormalBasis`.

YaelDillies commented:
I asked the people around me (Bhavik, Thomas Bloom, my PhD advisor), and we all agree that your generalisation should work, but also that we have no use for it. I am therefore leaving a TODO.

hrmacbeth commented:
I pushed a cheap generalization to inner product spaces (cheap because I kept the dimension restriction, so your original proof goes through without change).  OK?

hrmacbeth commented:
... at which point I figured I might as well do the general case.  Are you willing to keep it part of the PR?

YaelDillies commented:
Ahah, that's great! I was not willing to generalise because I have so many other things to push from APAP, but if you have already done it I won't complain 😅 Thanks a lot!

b-mehta commented on Mathlib/Analysis/InnerProductSpace/MarcinkiewiczZygmund.lean:
```diff
@@ -0,0 +1,286 @@
+  let b := stdOrthonormalBasis ℝ 𝕜; clear_value b
+  set B := A ^^ n
+  simp only [pow_mul, ← b.repr.norm_map, PiLp.norm_sq_eq_of_L2, map_sum, norm_eq_abs, sq_abs,
+    ge_iff_le, Fintype.sum_apply (γ := n)]
```
This looks like another case of `simp` giving `ge_iff_le` for no good reason

YaelDillies commented:
I have decided to move the new file to a new folder `Analysis.Discrete` because I have more discrete analysis in the making and Marcinkiewicz-Zygmund is usually classified as a discrete analysis result.

sgouezel commented on Mathlib/Analysis/Discrete/MarcinkiewiczZygmund.lean:
```diff
@@ -0,0 +1,286 @@
+/-!
+# The Marcinkiewicz-Zygmund inequality
+
+This file proves the Marcinkiewicz-Zygmund inequality. This is the statement that, for a random
```
I am really surprised by this version: for me, one of the important points of Marcinkiewicz-Zygmund is that you can take random variables which are not identically distributed. Also, you are working with the specific case of a fintype, while the inequality in true for a general probability space. I'd rather deduce the particular version from the general case (which is not really harder, although I'd rather get the more precise Rosenthal or Burkholder inequalities), but if you want to focus on the discrete space you should emphasize that in the file-level docstring.

b-mehta replied:
Indeed this is a special case, but it's precisely the special case which arises as a very early dependence in our work. Would you mind if we stayed with this version for now, with a comment to generalise it in the way you describe?

sgouezel replied:
I absolutely don't mind, as long as it's explained clearly in the file-level docstring.

YaelDillies replied:
Tried adding something

Parcly-Taxel commented on Mathlib/Analysis/Discrete/MarcinkiewiczZygmund.lean:
```diff
@@ -0,0 +1,288 @@
+* We currently only prove the inequality for `p = 2 * m` an even natural number. The general `p`
+  case can be obtained from this specific one by nesting of Lp norms.
+* We currently only prove the case of a uniformly distributed discrete random variable. This needs
+  to be generalised! (and when it is, this file should move out of the `Algebra.Discrete`) folder.
```
```suggestion
  to be generalised! (and when it is, this file should move out of the `Analysis.Discrete`) folder.
```

sgouezel commented on Mathlib/Analysis/Discrete/MarcinkiewiczZygmund.lean:
```diff
@@ -0,0 +1,290 @@
+local notation:70 s:70 " ^^ " n:71 => Fintype.piFinset fun _ : n ↦ s
+
+private lemma step_one (hA : A.Nonempty) (f : α → ℝ) (a : n → α)
+    (hf : ∀ i, ∑ a in A ^^ n, f (a i) = 0) :
```
```suggestion
    (hf : ∀ i, ∑ a ∈ A ^^ n, f (a i) = 0) :
```
The `in` notation is deprecated, can you use `∈` instead throughout?

sgouezel commented on Mathlib/Analysis/Discrete/MarcinkiewiczZygmund.lean:
```diff
@@ -0,0 +1,290 @@
+
+namespace Real
+
+/-- The **Marcinkiewicz-Zygmund inequality** for real-valued functions, with a slightly better
```
Can you spell out what the inequality says, in layman terms -- in particular since there is no explicit independence in the statement, while it's clearly the intuition to have in mind? Same thing for the other main theorems.

sgouezel commented on Mathlib/Analysis/Discrete/MarcinkiewiczZygmund.lean:
```diff
@@ -0,0 +1,290 @@
+
+Note that `InnerProductSpace.marcinkiewicz_zygmund` and `RCLike.marcinkiewicz_zygmund` are other
+versions that works for more general target spaces, at the expense of a slightly worse constant. -/
+theorem marcinkiewicz_zygmund (hm : m ≠ 0) (f : α → ℝ) (hf : ∀ i, ∑ a in A ^^ n, f (a i) = 0) :
```
Isn't it also true for `m = 0`? Same thing below.

sgouezel replied:
No, it's not, when `n` is empty. Sorry for the noise.

sgouezel commented on Mathlib/Analysis/Discrete/MarcinkiewiczZygmund.lean:
```diff
@@ -0,0 +1,290 @@
+open Finset Fintype Nat Real
+variable {α : Type*} {n : Type*} [Fintype n] {A : Finset α} {m : ℕ}
+
+private lemma step_seven {f : α → ℝ} {a b : n → α} :
```
To help your reader follow the argument, it would help a lot to have a short proof sketch in the file-level docstring, mentioning where the various auxiliary lemmas are relevant.

sgouezel commented on Mathlib/Analysis/InnerProductSpace/Basic.lean:
```diff
@@ -2257,6 +2257,27 @@ instance : InnerProductSpace ℝ ℂ := InnerProductSpace.complexToReal
+This is different from the inner product space structure obtained from
+`InnerProductSpace.rclikeToReal 𝕜 𝕜` because of a diamond in the `ℝ`-normed space instances:
+`NormedSpace.restrictScalars ℝ 𝕜 𝕜` ≠ `NormedAlgebra.toNormedSpace' (𝕜 := ℝ) (𝕜' := 𝕜)`. -/
+noncomputable def RCLike.innerProductSpaceReal : InnerProductSpace ℝ 𝕜 :=
```
can you use `where` instead?

YaelDillies committed:
feat: Marcinkiewicz-Zygmund inequality
Prove the Marcinkiewicz-Zygmund inequality.

The Marcinkiewicz-Zygmund inequality states that, if `X₁, ... Xₐ ∈ L^p` are independent random variables of mean zero valued in some inner product space, then the `L^p`-norm of `X₁ + ... + Xₐ` is at most `Cₚ` times the `L^(p/2)`-norm of `|X₁|² + ... + |Xₐ|²`, where `Cₚ` is a constant depending on `p` alone.

YaelDillies commented on Mathlib/Analysis/Discrete/MarcinkiewiczZygmund.lean:
```diff
@@ -0,0 +1,290 @@
+
+This file proves the Marcinkiewicz-Zygmund inequality. This is the statement that, for a random
+variable `f` valued in a real inner product space, the $$L^p$$-norm of the sum of `n` samples of `f`
+is bounded by the $$L^p$$-norm of `f` up to some multiplicative constant.
```
I am not sure I understand this quibble. Isn't the point to get *any* constant? At least, [Wikipedia](https://en.wikipedia.org/wiki/Marcinkiewicz%E2%80%93Zygmund_inequality) only claims the result with a constant, without specifying anything about its size and relationship to the central limit theorem.

b-mehta replied:
I think the quibble here is exactly the opposite! The usual MZ inequality allows any constant, but in our specific application the shape of the constant is actually important. And so what LeanAPAP needs as "the MZ inequality" is a bit different from the usual formulation, since we require better control on the constants. And then the docstring should say a bit more about what the control needed is

b-mehta replied:
(imo, your new docstring satisfies this)

YaelDillies commented:
Thank you Heather and Sébastien for the reviews. I have decided that the current proof was too ugly and specific to be worth spending more time on it, so I have started formalising the general random variable version.

Along the way, I've made another big change: Instead of generalising to inner product spaces by taking a basis (bootstrapping), I rewrote the entire proof to handle products of inner products rather than products. This makes the proof a bit more complicated in places, but overall seems to work nicer AND gives a constant that does not depend on the dimension of the space.

There are still a few sorries left (10 at the time of writing), but none of them should be particularly challenging.

YaelDillies commented on Mathlib/Probability/MarcinkiewiczZygmund.lean:
```diff
@@ -0,0 +1,187 @@
+    ∫ ω, ‖∑ i ∈ A, X i ω‖ ^ (2 * m) ∂μ ≤
+      marcinkiewiczZygmundSymmConst (2 * m) * ∫ ω, (∑ i ∈ A, ‖X i ω‖ ^ 2) ^ m ∂μ := by
+  have : DecidableEq ι := Classical.decEq _
+  -- Turn the `L^p` assumption on the `X i` into various integrability conditions.
+  have integrable_prod_norm_X I (hI : I ∈ A ×ˢ A ^^ m) :
+    Integrable (fun ω ↦ ∏ k, ‖X (I k).1 ω‖ * ‖X (I k).2 ω‖) μ := sorry
+  have integrable_prod_inner_X I (hI : I ∈ A ×ˢ A ^^ m) :
+    Integrable (fun ω ↦ ∏ k, inner (𝕜 := ℝ) (X (I k).1 ω) (X (I k).2 ω)) μ := sorry
```
@sgouezel, any idea for how to prove these integrability conditions swiftly? It's hopefully a simple application of Hölder

YaelDillies commented:
I will reopen this once I am done generalising to arbitrary measures.
